Refactored new ToolInput/Outputs so they reside in own packages and not in original interfaces

judovana opened the pull request:
This PR is refactoring the long interfaces with many inner classes to logical packages.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Leonid Kuskov](https://openjdk.org/census#lkuskov) (@lkuskov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/asmtools pull/53/head:pull/53` \
`$ git checkout pull/53`

Update a local copy of the PR: \
`$ git checkout pull/53` \
`$ git pull https://git.openjdk.org/asmtools pull/53/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 53`

View PR using the GUI difftool: \
`$ git pr show -t 53`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/asmtools/pull/53.diff">https://git.openjdk.org/asmtools/pull/53.diff</a>

</details>

judovana commented:
As promised earlier. I'm still hesitating with the usage of uEscWriter in string based inputs/outputs. Maybe do you have jasm/jdis class reproducer of possible issues?

jonathan-gibbons commented:
@judovana How much is it useful is refactor the code like this?   Some of the refactored classes are very small.  How much do the subtypes have subtype-specific API that needs to be directly and publicly available, compared to methods that just implement or override behavior declared in the super type.

Would it be better to just hide the subtypes where possible, and have static factory methods in the supertypes, possibly setting up the use of sealed classes and `permits` down the road?

judovana commented:
> @judovana How much is it useful is refactor the code like this? Some of the refactored classes are very small. How much 

They were originally placed in only two classes - in the top level directories themselves, as the number of resulting types was underestimated. The interface' java files had 500 lines+ and 7+ named inner clasess, where each was necessary to use for different usecase. 
I myself, as author of those classes, had to look half year later to the interface, and search for the implementation names. As they are now, you look into pakcage, and directly knows.

>do the subtypes have subtype-specific API that needs to be directly and publicly available, compared to methods that just implement or override behavior declared in the super type.

Thy do. Mos tof the implementations is 50-100 lines, and overwirte the mai work methods -  getDataInputStream for input and startclass/finshclass  for outputs.

> 
> Would it be better to just hide the subtypes where possible, and have static factory methods in the supertypes, possibly setting up the use of sealed classes and `permits` down the road?

Maybe. But I woudl rather have it opened as much as possible.
The main asmtool methods are allowng you to use any implementation of ToolInput/Output/Log which you desire.
The asmtools application itself, is using stdin/printout or  filein/dirout. Usage if asmtools as libraryand thus base for future IDE (and asmtools' unittests)  is mostly going to use  stringin/streamin/bytein and stringout/byteout 

I think the freedom to deal with input/output in flexible way is crucial for asmtool to have more chances to catch up as more widely used language/library, as it already now allowed to be more easy to use as cli tool.

lkuskov commented on src/org/openjdk/asmtools/common/inputs/ByteInput.java:
```diff
@@ -0,0 +1,66 @@
+package org.openjdk.asmtools.common.inputs;
```
Copyright should be added here.

lkuskov commented on src/org/openjdk/asmtools/common/inputs/ByteInput.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.Collection;
+import java.util.Optional;
+
+public class ByteInput implements ToolInput {
```
It would be good to have a class descriptor here

lkuskov commented on src/org/openjdk/asmtools/common/inputs/ByteInput.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public Collection<String> readAllLines() throws IOException {
+        init();
+        ArrayList r = new ArrayList();
```
Please do use self descriptive names for local vars not just one symbols (except loop's indices)

lkuskov commented on src/org/openjdk/asmtools/common/outputs/ByteOutput.java:
```diff
@@ -0,0 +1,100 @@
+    @Override
+    public void printlns(String line) {
+        try {
+            currentClass.write((line + "\n").getBytes(StandardCharsets.UTF_8));
```
Much better to use ` + System.lineSeparator()`

lkuskov commented:
Looks good to me. Please make cosmetic changes commented in diff like copyrings, code comments etc. and PR will be merged. Thanks

jonathan-gibbons commented on src/org/openjdk/asmtools/common/outputs/DirOutput.java:
```diff
@@ -0,0 +1,98 @@
+            outfile = new File(fqn + fileExtension.orElseGet(() -> ""));
+        } else {
+            environment.traceln("writing -d " + dir.getPath());
+            if (!fileSeparator.equals("/")) {
```
Minor implementation detail: the implementation of `replace` is a no-op if the two arguments are single character strings and equal.  So, there is no need for the external guard in the `if` statement.

lkuskov commented:
Hi Jiří,
Just curious, I've written an unit test for the PR [#54](https://github.com/openjdk/asmtools/pull/54) that uses your approach to test output.
```
        int rc = 0;
        if (files.isEmpty())
            fail(toolName + ": no files");
        List<String> toolArgs = new ArrayList<>();
        toolArgs.add("-d");
        toolArgs.add(destDir.getPath());
        toolArgs.addAll(files);
        ToolOutput.ByteOutput encodedFiles = new ToolOutput.ByteOutput();
        ToolOutput.StringLog encodeLog = new ToolOutput.StringLog();
        org.openjdk.asmtools.jasm.Main jasm = new org.openjdk.asmtools.jasm.Main(encodedFiles, encodeLog, toolArgs.toArray(new String[0]));
            jasm.setVerboseFlag(true);
            jasm.setTraceFlag(true);
            rc = jasm.compile();
            return new InputOutputTests.LogAndBinResults(encodedFiles, encodeLog, rc);
```
My jasm file warns:
jasm   -  WARN: test_1()V - The instruction 'ifge' uses the argument 0x8000 that exceeds allowed signed 16-bit value 0x7FFF
1 warning(s)

I need to parse this output to confirm that jasm works as expected. How can it be done? The encodeLog is empty.  
Thanks,
Leonid
PS It would be great if you become a  contributor. At least it will allow me assign you  a reviewer for Asmtools

judovana commented:
Hello!

I already have some powers: https://openjdk.org/census#jvanek  but not as much as you would need.

Yours test looks correct. Where the warning is printed? It seems that I have to miss some log somewhere, and so it is still writing to stderr or missing flush. After you call  `jasm.compile()`, you shoudl be done and `encodeLog` populated and all streasm empty.

Maybe you can add it to that PR and I can pull and try?

lkuskov commented:
Hello,
Since you are already committer/author of the jdk projects let's get started from the beginning for Code Tools. Please write a request as described in [Becoming an Author](https://openjdk.org/projects/#project-author). After you make 8 total contributions, you will be eligible to be [nominated to be a 'committer'](http://openjdk.java.net/projects/#project-committer).

I've uploaded my unit test, I fixed it and it works right now. Please review PR [#54](https://github.com/openjdk/asmtools/pull/54) I will push it it at8 branch

Thanks

judovana commented:
hello. Emialed the author claim, fixed hopefully what was rebuked. Will look into https://github.com/openjdk/asmtools/pull/54 tomorrow. Can you please highlight, what changed for "I fixed it and it works right now" please?

lkuskov commented:
LGTM Thanks
